File: java/app/src/main/java/com/huawei/hms/site/sample/SearchFragmentActivity.java
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
 */

package com.huawei.hms.site.sample;

import java.util.Arrays;
import java.util.List;

import com.google.gson.Gson;
import com.huawei.hms.site.api.model.AddressDetail;
import com.huawei.hms.site.api.model.Coordinate;
import com.huawei.hms.site.api.model.CoordinateBounds;
import com.huawei.hms.site.api.model.LocationType;
import com.huawei.hms.site.api.model.Poi;
import com.huawei.hms.site.api.model.SearchStatus;
import com.huawei.hms.site.api.model.Site;
import com.huawei.hms.site.widget.SearchFilter;
import com.huawei.hms.site.widget.SearchFragment;
import com.huawei.hms.site.widget.SiteSelectionListener;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

public class SearchFragmentActivity extends AppCompatActivity {
    private static final String TAG = "SearchFragmentActivity";

    private TextView textView;

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search_fragment);
        textView = findViewById(R.id.demo_fragment_textView_result);
        SearchFilter searchFilter = buildSearchFilter();
        if (null == searchFilter) {
            Log.e(TAG, "searchFilter is null.");
            return;
        }
        initFragment(searchFilter);
    }

    private void initFragment(SearchFilter searchFilter) {
        SearchFragment searchFragment =
            (SearchFragment) getSupportFragmentManager().findFragmentById(R.id.demo_fragment);
        searchFragment.setSearchFilter(searchFilter);
        searchFragment.setHint("Click to start the query.");
        String apiKey = Utils.getApiKey();
        searchFragment.setApiKey(apiKey);

        if (searchFragment != null) {
            searchFragment.setOnSiteSelectedListener(new SiteSelectionListener() {
                @Override
                public void onSiteSelected(Site data) {
                    Site site = data;

                    StringBuilder response = new StringBuilder("\n");
                    AddressDetail addressDetail = site.getAddress();
                    Coordinate location = site.getLocation();
                    Poi poi = site.getPoi();
                    CoordinateBounds viewport = site.getViewport();
                    response.append(String.format(
                        "siteId: '%s', Name: %s, FormatAddress: %s, Country: %s, CountryCode: %s, Location: %s, PoiTypes: %s, Viewport: %s, streetNumber: %s, postalCode: %s , tertiaryAdminArea: %s \n\n",
                        site.getSiteId(), site.getName(), site.getFormatAddress(),
                        (addressDetail == null ? "" : addressDetail.getCountry()),
                        (addressDetail == null ? "" : addressDetail.getCountryCode()),
                        (location == null ? "" : (location.getLat() + "," + location.getLng())),
                        (poi == null ? "" : Arrays.toString(poi.getPoiTypes())),
                        (viewport == null ? "" : viewport.getNortheast() + "," + viewport.getSouthwest()),
                        (addressDetail == null ? "" : addressDetail.getStreetNumber()),
                        (addressDetail == null ? "" : addressDetail.getPostalCode()),
                        (addressDetail == null ? "" : addressDetail.getTertiaryAdminArea())));
                    if ((poi != null)) {
                        Gson g = new Gson();
                        String jsonString = g.toJson(poi.getChildrenNodes());
                        response.append(String.format("childrenNode: %s,", jsonString));
                    }
                    textView.setText(response.toString());
                }

                @Override
                public void onError(SearchStatus status) {
                    textView.setText(" Error code from fragment: " + status.getErrorCode()
                        + "\n Error message from fragment: " + status.getErrorMessage());
                }
            });
        }
    }

    private SearchFilter buildSearchFilter() {
        SearchFilter searchFilter = null;
        if (null != getIntent()) {
            Intent intent = getIntent();

            String query = intent.getStringExtra("query");
            String language = intent.getStringExtra("language");
            String countryCode = intent.getStringExtra("countryCode");
            String radius = intent.getStringExtra("radius");
            CoordinateBounds bounds = intent.getParcelableExtra("bounds");
            Coordinate location = intent.getParcelableExtra("location");
            List<LocationType> poiTypes = (List<LocationType>) intent.getSerializableExtra("poiTypes");
            boolean children = intent.getBooleanExtra("children", false);
            boolean strictBounds = intent.getBooleanExtra("strictBounds", false);
            List<String> countryList = (List<String>) intent.getSerializableExtra("countries");

            searchFilter = new SearchFilter();

            if (!TextUtils.isEmpty(language)) {
                searchFilter.setLanguage(language);
            }
            if (!TextUtils.isEmpty(query)) {
                searchFilter.setQuery(query);
            }
            if (!TextUtils.isEmpty(countryCode)) {
                searchFilter.setCountryCode(countryCode);
            }
            if (null != countryList && countryList.size() > 0) {
                searchFilter.setCountries(countryList);
            }

            Integer radiusValue;
            if ((radiusValue = Utils.parseInt(radius)) != null) {
                searchFilter.setRadius(radiusValue);
            }

            if (null != bounds) {
                searchFilter.setBounds(bounds);
            }

            if (null != location) {
                searchFilter.setLocation(location);
            }

            if (null != poiTypes) {
                searchFilter.setPoiType(poiTypes);
            }
            searchFilter.setChildren(children);
            searchFilter.setStrictBounds(strictBounds);
        }
        return searchFilter;
    }
}

File: kotlin/app/src/main/java/com/huawei/hms/site/sample/SearchFragmentActivity.kt
/*
 * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
 */
package com.huawei.hms.site.sample

import android.os.Bundle
import android.util.Log
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.google.gson.Gson
import com.huawei.hms.site.api.model.*
import com.huawei.hms.site.sample.Utils.apiKey
import com.huawei.hms.site.sample.Utils.parseInt
import com.huawei.hms.site.widget.SearchFilter
import com.huawei.hms.site.widget.SearchFragment
import com.huawei.hms.site.widget.SiteSelectionListener
import java.util.*

class SearchFragmentActivity : AppCompatActivity() {
    companion object {
        private const val TAG = "SearchFragmentActivity"
    }

    private lateinit var textView: TextView
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search_fragment)
        textView = findViewById(R.id.demo_fragment_textView_result)
        val searchFilter = buildSearchFilter()
        if (null == searchFilter) {
            Log.e(TAG, "searchFilter is null.")
            return
        }
        initFragment(searchFilter)
    }

    private fun initFragment(searchFilter: SearchFilter) {
        val searchFragment = supportFragmentManager.findFragmentById(R.id.demo_fragment) as SearchFragment?
        searchFragment?.setSearchFilter(searchFilter)
        searchFragment?.setHint("Click to start the query.")
        val apiKey = apiKey
        searchFragment?.setApiKey(apiKey)
        searchFragment?.setOnSiteSelectedListener(object : SiteSelectionListener {
            override fun onSiteSelected(data: Site) {
                val response = StringBuilder("\n")
                val addressDetail = data.getAddress()
                val location = data.getLocation()
                val poi = data.getPoi()
                val viewport = data.getViewport()
                response.append(String.format(
                        "siteId: '%s', Name: %s, FormatAddress: %s, Country: %s, CountryCode: %s, Location: %s, PoiTypes: %s, Viewport: %s, streetNumber: %s, postalCode: %s , tertiaryAdminArea: %s \n\n",
                        data.getSiteId(), data.getName(), data.getFormatAddress(),
                        if (addressDetail == null) "" else addressDetail.getCountry(),
                        if (addressDetail == null) "" else addressDetail.getCountryCode(),
                        if (location == null) "" else location.getLat().toString() + "," + location.getLng(),
                        if (poi == null) "" else Arrays.toString(poi.getPoiTypes()),
                        if (viewport == null) "" else viewport.getNortheast().toString() + "," + viewport.getSouthwest(),
                        if (addressDetail == null) "" else addressDetail.getStreetNumber(),
                        if (addressDetail == null) "" else addressDetail.getPostalCode(),
                        if (addressDetail == null) "" else addressDetail.getTertiaryAdminArea()))
                if (poi != null) {
                    val g = Gson()
                    val jsonString = g.toJson(poi.getChildrenNodes())
                    response.append(String.format("childrenNode: %s,", jsonString))
                }
                textView.text = response.toString()
            }

            override fun onError(status: SearchStatus) {
                textView.text = "Error code from fragment: ${status.getErrorCode()}  Error message from fragment: ${status.getErrorMessage()}"
            }
        })
    }

    private fun buildSearchFilter(): SearchFilter? {
        var searchFilter: SearchFilter? = null
        if (null != intent) {
            val intent = intent
            val query: String? = intent.getStringExtra("query")
            val language: String? = intent.getStringExtra("language")
            val countryCode: String? = intent.getStringExtra("countryCode")
            val radius: String? = intent.getStringExtra("radius")
            val bounds: CoordinateBounds? = intent.getParcelableExtra("bounds")
            val location: Coordinate? = intent.getParcelableExtra("location")
            var poiTypes: List<LocationType>? = null
            if (null != intent.getSerializableExtra("poiTypes")) {
                poiTypes = intent.getSerializableExtra("poiTypes") as List<LocationType>
            }
            val children: Boolean? = intent.getBooleanExtra("children", false)
            val strictBounds: Boolean? = intent.getBooleanExtra("strictBounds", false)
            searchFilter = SearchFilter()
            if (!language.isNullOrEmpty()) {
                searchFilter.language = language
            }
            if (!query.isNullOrEmpty()) {
                searchFilter.query = query
            }
            if (!countryCode.isNullOrEmpty()) {
                searchFilter.countryCode = countryCode
            }
            if (!radius.isNullOrEmpty()) {
                val radiusValue: Int? = parseInt(radius!!)
                searchFilter.radius = radiusValue
            }
            if (null != bounds) {
                searchFilter.bounds = bounds
            }
            if (null != location) {
                searchFilter.location = location
            }
            if (null != poiTypes) {
                searchFilter.poiType = poiTypes
            }
            searchFilter.isChildren = children!!
            searchFilter.strictBounds = strictBounds
        }
        return searchFilter
    }
}